File: typescript/src/cache.ts
import type { ProviderRef } from "./providers/types.js";

export class UploadCache {
  private entries = new Map<string, ProviderRef>();

  constructor(private clock: () => number = () => Date.now() / 1000) {}

  get(providerName: string, contentHash: string): ProviderRef | null {
    const key = `${providerName}:${contentHash}`;
    const ref = this.entries.get(key);
    if (!ref) return null;
    if (ref.expiresAt !== null && ref.expiresAt <= this.clock()) {
      this.entries.delete(key);
      return null;
    }
    return ref;
  }

  put(providerName: string, contentHash: string, ref: ProviderRef): void {
    this.entries.set(`${providerName}:${contentHash}`, ref);
  }
}

File: python/src/siphon/cache.py
"""In-memory, expiry-aware cache of provider upload references."""

from __future__ import annotations

import time
from typing import Callable

from siphon.providers import ProviderRef


class UploadCache:
    def __init__(self, clock: Callable[[], float] = time.time) -> None:
        self._clock = clock
        self._entries: dict[tuple[str, str], ProviderRef] = {}

    def get(self, provider_name: str, content_hash: str) -> ProviderRef | None:
        key = (provider_name, content_hash)
        ref = self._entries.get(key)
        if ref is None:
            return None
        if ref.expires_at is not None and ref.expires_at <= self._clock():
            del self._entries[key]
            return None
        return ref

    def put(self, provider_name: str, content_hash: str, ref: ProviderRef) -> None:
        self._entries[(provider_name, content_hash)] = ref
